### PR TITLE
Add integration tests

### LINKED ISSUE #11: Create integration tests
Should probably just use Node.js with `chai-http` to save time.

### PR DESCRIPTION
**Pull request by igorveselinovic**
Resolves #11 

Just wrote a couple tests for all the `karen` API endpoints. Should be able to see them running (and hopefully passing) in the pipeline.

Wrote a setup that runs **before** each test that:
- creates a pre-existing user that can be used by the test

Wrote a teardown that runs **after** each test that:
- deletes any created users (including the pre-existing user from the setup)

Also, wrote a `wait.js` script that just pings the given host until it returns a 404, so the CI pipeline runs that first, then runs the tests after.

Had to fix a few bugs:
- The selected database doesn't persist across connections, so we need to re-open the DB after creating the `karen` database.
- `DELETE` and `PATCH` both didn't return 404s when the user didn't exist, so fixed that by checking the rows affected count.
- The `auth` endpoint was returning 500 when it should've returned 404 because it wasn't checking for the right error.
- The `GET` endpoint shouldn't return an ID.
- The `"avatar_url"` field should be returned even when it's an empty string, and `PATCH` should support setting the `"avatar_url"` to an empty string.

### CODE REVIEW

File: models/user.go [hunk 1]
```diff
@@ -38,16 +38,21 @@ func (db *DB) CheckUser(user *User) (*User, error) {
 	return userFromDB, err
 }
 
-func (db *DB) DeleteUser(userID int64) error {
+// DeleteUser removesa a row from the "users" table.
```
Comment (Thao-Tran) [inline models/user.go]:
```suggestion
// DeleteUser removes a row from the "users" table.
```

File: test/test.js [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+    const res = await chai.request(karenEndpoint)
+      .post('/karen/v1/users')
+      .send({
+        name: 'Foo Bar',
+        email: 'foo@bar.baz',
+        password: 'foobarbaz',
+      });
+
+    expect(res).to.have.status(201);
+    expect(res.body).to.have.keys('id', 'name', 'email');
```
Comment (Thao-Tran) [inline test/test.js]: The chai docs recommends using `all.keys` even though just `keys` defaults to `all.keys` just to help with readability.
```suggestion
    expect(res.body).to.have.all.keys('id', 'name', 'email');
```

File: test/test.js [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+});
+
+describe('GET /karen/v1/users/self', () => {
+  it('should retrieve the session user', async () => {
+    const res = await chai.request(karenEndpoint)
+      .get('/karen/v1/users/self')
+      .set('User-ID', preExistingUser.id);
+
+    expect(res).to.have.status(200);
+    expect(res.body).to.have.keys('id', 'name', 'email');
```
Comment (Thao-Tran) [inline test/test.js]:
```suggestion
    expect(res.body).to.have.all.keys('id', 'name', 'email');
```

File: test/test.js [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+  });
+
+  it('should retrieve the session user with specified fields', async () => {
+    const res = await chai.request(karenEndpoint)
+      .get('/karen/v1/users/self')
+      .query({ includes: ['name', 'email'] })
+      .set('User-ID', preExistingUser.id);
+
+    expect(res).to.have.status(200);
+    expect(res.body).to.have.keys('name', 'email');
```
Comment (Thao-Tran) [inline test/test.js]:
```suggestion
    expect(res.body).to.have.all.keys('name', 'email');
```

File: test/test.js [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+});
+
+describe('GET /karen/v1/users/{user_id}', () => {
+  it('should retrieve the specified user', async () => {
+    const res = await chai.request(karenEndpoint)
+      .get(`/karen/v1/users/${preExistingUser.id}`)
+      .set('User-ID', 123);
+
+    expect(res).to.have.status(200);
+    expect(res.body).to.have.keys('id', 'name', 'email');
```
Comment (Thao-Tran) [inline test/test.js]:
```suggestion
    expect(res.body).to.have.all.keys('id', 'name', 'email');
```

File: test/test.js [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+      .set('User-ID', preExistingUser.id)
+      .send({
+        name: 'New Name',
+        email: 'newemail@foo.bar',
+        password: 'newpassowrd',
+        avatar_url: 'example.com/newavatar.png',
+      });
+
+    expect(res).to.have.status(200);
+    expect(res.body).to.have.keys('name', 'email', 'avatar_url');
```
Comment (Thao-Tran) [inline test/test.js]:
```suggestion
    expect(res.body).to.have.all.keys('name', 'email', 'avatar_url');
```

File: test/test.js [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+
+describe('PATCH /karen/v1/users/self', () => {
+  it('should update a user', async () => {
+    const res = await chai.request(karenEndpoint)
+      .patch('/karen/v1/users/self')
+      .set('User-ID', preExistingUser.id)
+      .send({
+        name: 'New Name',
+        email: 'newemail@foo.bar',
+        password: 'newpassowrd',
```
Comment (Thao-Tran) [inline test/test.js]: Haha did you mean to spell `password` wrong?

File: test/test.js [hunk 8]
```diff
@@ -0,0 +1,276 @@
...
+        avatar_url: 'example.com/newavatar.png',
+      });
+
```
Comment (Thao-Tran) [inline test/test.js]: Should maybe add a GET call to check that the user was updated properly.

File: test/test.js [hunk 9]
```diff
@@ -0,0 +1,276 @@
...
+  });
+
+  it('should update only the specified field(s)', async () => {
+    const res = await chai.request(karenEndpoint)
+      .patch('/karen/v1/users/self')
+      .set('User-ID', preExistingUser.id)
+      .send({
+        name: 'New Name',
+      });
+
```
Comment (Thao-Tran) [inline test/test.js]: Same as above.

File: test/test.js [hunk 10]
```diff
@@ -0,0 +1,276 @@
...
+    expect(res).to.have.status(200);
+    expect(res.body).to.have.keys('name');
```
Comment (Thao-Tran) [inline test/test.js]:
```suggestion
    expect(res.body).to.have.all.keys('name');
```

File: test/test.js [hunk 11]
```diff
@@ -0,0 +1,276 @@
...
+  it('should authenticate correct credentials', async () => {
+    const res = await chai.request(karenEndpoint)
+      .post('/karen/v1/users/auth')
+      .send({
+        email: preExistingUser.email,
+        password: preExistingUser.password,
+      });
+
+    expect(res).to.have.status(200);
+    expect(res.body).to.have.keys('id', 'name', 'email');
```
Comment (Thao-Tran) [inline test/test.js]:
```suggestion
    expect(res.body).to.have.all.keys('id', 'name', 'email');
```

File: test/test.js (hunk 7 above)
Comment (igorveselinovic) [reply test/test.js]: No...

File: test/test.js (hunk 8 above)
Comment (igorveselinovic) [reply test/test.js]: Yeah, I guess so. But then we should probably also add a GET call in the POST and DELETE tests that modify the database too, right?

Comment (Thao-Tran) [reply test/test.js]: Yeah definitely. I think I just forgot haha.

File: test/test.js (hunk 2 above)
Comment (igorveselinovic) [reply test/test.js]: Yeah, forgot about that.

Comment (Thao-Tran) [reply test/test.js]: I just know because I had to go to the docs to check how the `keys` thing works haha.

File: models/user.go [hunk 12]
```diff
@@ -10,24 +10,62 @@ import (
...
+	return newUser
+}
+
 // CheckUser authenticates the user's credentials and returns the user.
 func (db *DB) CheckUser(user *User) (*User, error) {
 	userFromDB := &User{}
-	queryString := fmt.Sprintf("SELECT ID, Name, Password FROM %s WHERE EMAIL=?", usersTable)
-	err := db.QueryRow(queryString, user.Email).Scan(&(userFromDB.ID),
-		&(userFromDB.Name), &(userFromDB.Password))
+	queryString := fmt.Sprintf("SELECT ID, Name, Email, Password, AvatarUrl FROM %s WHERE EMAIL=?", usersTable)
```
Comment (Thao-Tran) [inline models/user.go]: Is the idea now that `heimdall` is going to return these fields to the client when it hits the `/heimdall/v1/token` route, i.e. signs in?

Comment (igorveselinovic) [reply models/user.go]: Um, no... I think I was just on autopilot and didn't think about it. I was just treating it like a regular `POST` essentially. Do you think the handler should just return `"id"`, `"name"`, and `"email"`, then?

Comment (Thao-Tran) [reply models/user.go]: Oh haha yeah I think so because that's all `heimdall` actually needs.

**Pull request merged**